lib/logflare/backends/adaptor/postgres_adaptor/shared_repo.ex

```diff
@@ -0,0 +1,97 @@
+defmodule Logflare.Backends.Adaptor.PostgresAdaptor.SharedRepo do
+  use Ecto.Repo,
+    otp_app: :logflare,
+    adapter: Ecto.Adapters.Postgres
+
+  require Logger
+
+  alias Logflare.Source
+  alias Logflare.Backends.Backend
+
+  alias Logflare.Backends.Adaptor.PostgresAdaptor.Supervisor
+  alias Logflare.Backends.Adaptor.PostgresAdaptor.Repo.Migrations
+
+  def start(%Backend{} = backend) do
+    config = backend.config
+    pool_size = Application.get_env(:logflare, :postgres_backend_adapter)[:pool_size] || 10
+    schema = config["schema"] || config[:schema]
+
+    opts = [
+      url: config["url"] || config.url,
+      name: Supervisor.via(backend),
+      pool_size: pool_size,
+      # Wait until repo is fully up and running
+      sync_connect: true,
+      after_connect: {__MODULE__, :__after_connect__, [schema]}
+    ]
+
+    case Supervisor.start_child({__MODULE__, opts}) do
+      {:ok, pid} -> {:ok, pid}
+      # If there is already started process for this backend, then simply reuse
+      # it
+      {:error, {:already_started, pid}} -> {:ok, pid}
+      other -> other
+    end
+  end
+
+  @moduledoc """
+  Shared repository for all connections in Postgres Adaptor
+  """
+
+  def set_repo(%Backend{} = backend) do
+    with {:ok, pid} <- start(backend) do
+      {:ok, put_dynamic_repo(pid)}
+    end
+  end
+
+  def with_repo(backend, func) do
+    {:ok, old_repo} = set_repo(backend)
+
+    try do
+      func.()
+    after
+      put_dynamic_repo(old_repo)
+    end
+  end
+
+  @doc """
+  Run migrations for currently selected repository
+  """
+  @spec migrate!(Source.t()) :: :ok | {:error, term()}
+  def migrate!(%Source{} = source) do
```

Comment:
if it won't error due to rescuing then shouldn't use bang func

Reply:
My idea for a bang there is that it is destructive operation that caller should avoid using and be aware of the consequences (not all banged functions throws on error, see `var!/1`). But I can change it.

Reply:
OK cool, adding some docs on what is executed would be good

Reply:
Yea sound reasoning for the bang, let's keep it